File: 29Items-Projects/7Algorithms/7-String-algorithms/python/string_algorithms/algorithms/suffix_tree.py
from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class _Node:
    children: dict[str, "_Edge"] = field(default_factory=dict)
    terminal: bool = False


@dataclass
class _Edge:
    label: str
    target: _Node


class SuffixTree:
    def __init__(self, text: str) -> None:
        self._root = _Node()
        for start in range(len(text)):
            self._insert(text[start:])

    def contains(self, pattern: str) -> bool:
        current = self._root
        pattern_index = 0

        while pattern_index < len(pattern):
            edge = current.children.get(pattern[pattern_index])
            if edge is None:
                return False

            edge_index = 0
            while edge_index < len(edge.label) and pattern_index < len(pattern):
                if edge.label[edge_index] != pattern[pattern_index]:
                    return False
                edge_index += 1
                pattern_index += 1

            if pattern_index == len(pattern):
                return True

            current = edge.target
        return True

    @property
    def edge_count(self) -> int:
        return self._count_edges(self._root)

    def _insert(self, suffix: str) -> None:
        current = self._root
        remaining = suffix

        while remaining:
            edge = current.children.get(remaining[0])
            if edge is None:
                leaf = _Node(terminal=True)
                current.children[remaining[0]] = _Edge(remaining, leaf)
                return

            common_length = _common_prefix_length(remaining, edge.label)
            if common_length == len(edge.label):
                current = edge.target
                remaining = remaining[common_length:]
                continue

            self._split_edge(current, edge, common_length, remaining)
            return

        current.terminal = True

    def _split_edge(
        self, parent: _Node, edge: _Edge, common_length: int, remaining: str
    ) -> None:
        common_prefix = edge.label[:common_length]
        old_remainder = edge.label[common_length:]
        intermediate = _Node()
        intermediate.children[old_remainder[0]] = _Edge(old_remainder, edge.target)

        edge.label = common_prefix
        edge.target = intermediate
        parent.children[common_prefix[0]] = edge

        new_remainder = remaining[common_length:]
        if new_remainder:
            leaf = _Node(terminal=True)
            intermediate.children[new_remainder[0]] = _Edge(new_remainder, leaf)
        else:
            intermediate.terminal = True

    def _count_edges(self, node: _Node) -> int:
        return len(node.children) + sum(
            self._count_edges(edge.target) for edge in node.children.values()
        )


def _common_prefix_length(left: str, right: str) -> int:
    index = 0
    limit = min(len(left), len(right))
    while index < limit and left[index] == right[index]:
        index += 1
    return index

File: 29Items-Projects/7Algorithms/7-String-algorithms/src/main/java/com/stringalgorithms/algorithms/SuffixTree.java
package com.stringalgorithms.algorithms;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public final class SuffixTree {
    private final Node root = new Node();

    public SuffixTree(String text) {
        Objects.requireNonNull(text, "text must not be null");
        for (int start = 0; start < text.length(); start++) {
            insert(text.substring(start));
        }
    }

    public boolean contains(String pattern) {
        Objects.requireNonNull(pattern, "pattern must not be null");
        if (pattern.isEmpty()) {
            return true;
        }

        Node current = root;
        int patternIndex = 0;

        while (patternIndex < pattern.length()) {
            Edge edge = current.children.get(pattern.charAt(patternIndex));
            if (edge == null) {
                return false;
            }

            int edgeIndex = 0;
            while (edgeIndex < edge.label.length() && patternIndex < pattern.length()) {
                if (edge.label.charAt(edgeIndex) != pattern.charAt(patternIndex)) {
                    return false;
                }
                edgeIndex++;
                patternIndex++;
            }

            if (patternIndex == pattern.length()) {
                return true;
            }

            current = edge.target;
        }
        return true;
    }

    public int edgeCount() {
        return countEdges(root);
    }

    private void insert(String suffix) {
        Node current = root;
        String remaining = suffix;

        while (!remaining.isEmpty()) {
            Edge edge = current.children.get(remaining.charAt(0));
            if (edge == null) {
                Node leaf = new Node();
                leaf.terminal = true;
                current.children.put(remaining.charAt(0), new Edge(remaining, leaf));
                return;
            }

            int commonLength = commonPrefixLength(remaining, edge.label);
            if (commonLength == edge.label.length()) {
                current = edge.target;
                remaining = remaining.substring(commonLength);
                continue;
            }

            splitEdge(current, edge, commonLength, remaining);
            return;
        }

        current.terminal = true;
    }

    private void splitEdge(Node parent, Edge edge, int commonLength, String remaining) {
        String commonPrefix = edge.label.substring(0, commonLength);
        String oldRemainder = edge.label.substring(commonLength);
        Node intermediate = new Node();
        intermediate.children.put(oldRemainder.charAt(0), new Edge(oldRemainder, edge.target));

        edge.label = commonPrefix;
        edge.target = intermediate;
        parent.children.put(commonPrefix.charAt(0), edge);

        String newRemainder = remaining.substring(commonLength);
        if (newRemainder.isEmpty()) {
            intermediate.terminal = true;
        } else {
            Node leaf = new Node();
            leaf.terminal = true;
            intermediate.children.put(newRemainder.charAt(0), new Edge(newRemainder, leaf));
        }
    }

    private static int commonPrefixLength(String left, String right) {
        int limit = Math.min(left.length(), right.length());
        int index = 0;
        while (index < limit && left.charAt(index) == right.charAt(index)) {
            index++;
        }
        return index;
    }

    private static int countEdges(Node node) {
        int total = node.children.size();
        for (Edge edge : node.children.values()) {
            total += countEdges(edge.target);
        }
        return total;
    }

    private static final class Node {
        private final Map<Character, Edge> children = new HashMap<>();
        private boolean terminal;
    }

    private static final class Edge {
        private String label;
        private Node target;

        private Edge(String label, Node target) {
            this.label = label;
            this.target = target;
        }
    }
}
